### PR TITLE
[GEOS-8799] NULL namespace prefixes used in AppSchema WFS GetFeature responses

### PR DESCRIPTION
**Pull request by fernandor777**
Fix and tests for [GEOS-8799] "WFS GetFeature response contains NULL namespaces for App-Schema complex features when multiple queries are used".

https://osgeo-org.atlassian.net/browse/GEOS-8799

### CODE REVIEW

File: src/wfs/src/main/java/org/geoserver/wfs/xml/FeatureTypeSchemaBuilder.java [hunk 1]
```diff
@@ -384,6 +366,39 @@ public final XSDSchema buildSchemaInternal(
         return schema;
     }
 
+    private void addAllNamespacesFromCatalog(XSDSchema schema) {
```
Comment (nmco) [inline src/wfs/src/main/java/org/geoserver/wfs/xml/FeatureTypeSchemaBuilder.java]: Nice, avoiding repating code is always good karma :)

File: src/extension/app-schema/app-schema-test/src/test/java/org/geoserver/test/NamespacesWfsTest.java [hunk 2]
```diff
@@ -336,4 +346,80 @@ private void checkCount(
...
+                WFS20_XPATH_ENGINE,
+                document,
+                1,
+                "/wfs:FeatureCollection/wfs:member/wfs:FeatureCollection/wfs:member/"
+                        + "ms_gml32:Measurement_gml32");
+        // check prefixes:
+        String output = toString(document);
+        assertTrue(output.indexOf("null:Measurement_gml32") < 0);
+        assertTrue(output.indexOf("null:Station_gml32") < 0);
+        assertTrue(output.indexOf("ms_gml32:Measurement_gml32") > -1);
```
Comment (nmco) [inline src/extension/app-schema/app-schema-test/src/test/java/org/geoserver/test/NamespacesWfsTest.java]: Why -1 ? It should always be be greater than zero, no ?

File: src/extension/app-schema/app-schema-test/src/test/java/org/geoserver/test/NamespacesWfsTest.java [hunk 3]
```diff
@@ -336,4 +346,80 @@ private void checkCount(
...
+
+        checkCount(
+                WFS11_XPATH_ENGINE,
+                document,
+                1,
+                "/wfs:FeatureCollection/gml:featureMember/" + "ms_gml31:Measurement_gml31");
+        // check prefixes:
+        assertTrue(output.indexOf("null:Measurement_gml31") < 0);
+        assertTrue(output.indexOf("null:Station_gml31") < 0);
+        assertTrue(output.indexOf("ms_gml31:Measurement_gml31") > -1);
```
Comment (nmco) [inline src/extension/app-schema/app-schema-test/src/test/java/org/geoserver/test/NamespacesWfsTest.java]: Same as above.

**Comment by nmco** [comment]
Apart from some minor comments, looks good and its aligned with what has been done for the situation where we have only a single feature type queried.

File: src/wfs/src/main/java/org/geoserver/wfs/xml/FeatureTypeSchemaBuilder.java [hunk 4]
```diff
@@ -320,6 +300,8 @@ public final XSDSchema buildSchemaInternal(
                                     schemaURIs.get(namespace),
                                     imports,
                                     includes);
+                            // add prefix
+                            schema.getQNamePrefixToNamespaceMap().put(prefix, namespace);
```
Comment (nmco) [inline src/wfs/src/main/java/org/geoserver/wfs/xml/FeatureTypeSchemaBuilder.java]: Is this still needed ? If this is a complex feature type all the name-spaces should already have been added in the ``if`` above right ?

File: src/extension/app-schema/app-schema-test/src/test/java/org/geoserver/test/NamespacesWfsTest.java (hunk 2 above)
Comment (fernandor777) [reply src/extension/app-schema/app-schema-test/src/test/java/org/geoserver/test/NamespacesWfsTest.java]: String.indexOf returns the index of the first occurrence of the specified substring, or -1 if there is no such occurrence.  So if is == -1 substring not found.

Comment (nmco) [reply src/extension/app-schema/app-schema-test/src/test/java/org/geoserver/test/NamespacesWfsTest.java]: Ah! right my bad, I thought you were doing a XPATH count.

File: src/wfs/src/main/java/org/geoserver/wfs/xml/FeatureTypeSchemaBuilder.java (hunk 4 above)
Comment (fernandor777) [reply src/wfs/src/main/java/org/geoserver/wfs/xml/FeatureTypeSchemaBuilder.java]: Yes, it sounds sensible, deleting it.

File: src/extension/app-schema/app-schema-test/src/test/java/org/geoserver/test/NamespacesWfsTest.java (hunk 3 above)
Comment (fernandor777) [reply src/extension/app-schema/app-schema-test/src/test/java/org/geoserver/test/NamespacesWfsTest.java]: -1 == substring not found.

**Comment by nmco** [comment]
The Windows build failure looks unrelated to this PR:
```
[INFO] gs-restconfig ...................................... FAILURE [02:22 min]
```

... and the same build failure happen on other PRs. I'm merging this.

**Pull request merged**